### ISSUE TITLE
Rewrite code in Rust

### ISSUE THREAD

**Issue by denisidoro** [denisidoro/navi#154]
This project started as just a few lines of code then it gained some traction. More features were introduced. The code became cluttered. Even though I still have a clear vision of everything that is happening, I understand it is a little bit difficult for a newcomer to know where he/she should look at.

The namespaces don't make much sense anymore. In the beginning, everything that resulted into printing something into the screen was located in `ui.sh`. I took some shortcuts and now there are "model" namespaces such as `arg.sh` that call `fzf`, for example.

Considering I don't plan to write this in another programming language in the near future, I'd like to use a more robust design pattern. One that won't allow me to take shortcuts.

In a nutshell, I'd like the following: `ui = f(state)`.

Considering this is bash and there's nothing like react available, I'll need to scratch my head a little bit.

Any suggestions are welcome!

**Comment by denisidoro** [comment]
After much thought, I came to the conclusion that I won't be able to write a decent code in bash, regardless of the design pattern.

I intend to release a version 1.0 in bash and keep adding some bug fixes to it for some time.

In parallel, I'll start working on a new branch 2.0 rewriting the whole code in Rust, Go or V. Probably Rust.

**Comment by denisidoro** [comment]
Development has started! https://github.com/denisidoro/navi/tree/rust
Any help is appreciated!!!

**Comment by denisidoro** [comment]
If you want to contribute, please grab a task from here: https://github.com/denisidoro/navi/projects/1

**Comment by denisidoro** [comment]
Closing this because https://github.com/denisidoro/navi/pull/200 contains most features already. I'll open a separate issue dealing with missing features specifically.
